### PR TITLE
Add support for GitHub Enterprise IDP

### PR DESCRIPTION
**Pull request by enj**
This change adds two new fields to the GitHubIdentityProvider:

1. Hostname
2. CA

Both of these fields are optional.  When unspecified, the IDP behaves exactly the same and uses github.com as the backing identity store.  When hostname is specified, the IDP is reconfigured to use the GitHub Enterprise API endpoints with the given hostname.  The CA file may optionally be provided to validate the connections made to the GitHub Enterprise instance.

Unit tests were added to validate the new fields (along with the existing fields since they did not have any tests).

Signed-off-by: Monis Khan <mkhan@redhat.com>

---

https://trello.com/c/1zmyNYYi

/assign @liggitt @simo5

@openshift/sig-security

/hold

Holding for 3.11

### CODE REVIEW

**Comment by simo5** [comment]
I cannot test GHE but the changes are obvious and trivial
/lgtm

File: pkg/cmd/server/apis/config/validation/oauth.go [hunk 1]
```diff
@@ -354,9 +356,35 @@ func ValidateGitHubIdentityProvider(provider *configapi.GitHubIdentityProvider,
...
 	}
 
+	if hostname := provider.Hostname; len(hostname) != 0 {
+		hostnamePath := fieldPath.Child("hostname")
+
+		if hostname == "github.com" || strings.HasSuffix(hostname, ".github.com") {
+			validationResults.AddErrors(field.Invalid(hostnamePath, hostname, "cannot equal [*.]github.com"))
+		}
+
+		if !isValidHostname(hostname) {
```
Comment (liggitt) [inline pkg/cmd/server/apis/config/validation/oauth.go]: no scheme, no port?

Comment (enj) [reply pkg/cmd/server/apis/config/validation/oauth.go]: GHE does not let you configure that.  Https with no port only.

File: pkg/oauthserver/oauth/external/github/github.go [hunk 2]
```diff
@@ -82,17 +112,35 @@ func NewProvider(providerName, clientID, clientSecret string, organizations, tea
...
 	}
 
-	return &provider{
+	p := &provider{
 		providerName:         providerName,
 		clientID:             clientID,
 		clientSecret:         clientSecret,
 		allowedOrganizations: allowedOrganizations,
 		allowedTeams:         allowedTeams,
+		transport:            transport,
```
Comment (liggitt) [inline pkg/oauthserver/oauth/external/github/github.go]: only set transport in the hostname block below?

File: pkg/cmd/server/apis/config/validation/oauth.go (hunk 1 above)
Comment (enj) [reply pkg/cmd/server/apis/config/validation/oauth.go]: (based on the limited feedback the UI provides - the configuration settings are very limited)

File: pkg/oauthserver/oauth/external/github/github.go [hunk 3]
```diff
@@ -82,17 +112,35 @@ func NewProvider(providerName, clientID, clientSecret string, organizations, tea
...
+		githubAuthorizeURL:   defaultGithubAuthorizeURL,
+		githubTokenURL:       defaultGithubTokenURL,
+		githubUserApiURL:     defaultGithubUserApiURL,
+		githubUserOrgURL:     defaultGithubUserOrgURL,
+		githubUserTeamURL:    defaultGithubUserTeamURL,
+		githubUserEmailURL:   defaultGithubUserEmailURL,
+	}
+
+	if len(hostname) != 0 {
```
Comment (liggitt) [inline pkg/oauthserver/oauth/external/github/github.go]: maybe switch to

```
if len(hostname) > 0 {
... set custom endpoints ...
}else {
... set standard endpoints ...
}
```

that lessens the chance we end up with a standard endpoint left in the struct that should have been a custom one

File: pkg/oauthserver/oauth/external/github/github.go (hunk 2 above)
Comment (liggitt) [reply pkg/oauthserver/oauth/external/github/github.go]: ah, just got to where you construct it, and set it to DefaultTransport

File: pkg/oauthserver/oauth/external/github/github.go [hunk 4]
```diff
@@ -18,14 +18,29 @@ import (
...
+	enterpriseGithubUserOrgURL   = "https://%s/api/v3/user/orgs"
+	enterpriseGithubUserTeamURL  = "https://%s/api/v3/user/teams"
+	enterpriseGithubUserEmailURL = "https://%s/api/v3/user/emails"
+
+	// GitHub OAuth scopes, see provider.NewConfig
+	githubOAuthScope = "user:email"
+	githubOrgScope   = "read:org"
 
 	// https://developer.github.com/v3/#current-version
 	// https://developer.github.com/v3/media/#request-specific-version
```
Comment (liggitt) [inline pkg/oauthserver/oauth/external/github/github.go]: is the accept header the same for enterprise?

**Comment by liggitt** [comment]
a couple nits/suggestions, generated bits around the api server config types (deepcopy/defaults) need updating, as well as the types unit test

lgtm otherwise

File: pkg/oauthserver/oauth/external/github/github.go [hunk 5]
```diff
@@ -277,12 +325,16 @@ func page(initialURL, token string, newObj func() interface{}, processObj func(i
...
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token))
 	req.Header.Set("Accept", githubAccept)
 
-	res, err := http.DefaultClient.Do(req)
+	res, err := p.transport.RoundTrip(req)
```
[the remark this answers is not in the document]
Comment (enj) [reply pkg/oauthserver/oauth/external/github/github.go]: Client does redirects and cookies.  Neither seemed appropriate for a backend API call...

File: pkg/oauthserver/oauth/external/github/github.go (hunk 3 above)
Comment (enj) [reply pkg/oauthserver/oauth/external/github/github.go]: But that does not actually help to make sure you set it correctly.  The changes required for a new API are now:

1. create two new constants
1. create new field
1. set field in two different places

The unit test was a (probably futile) attempt to make the last step less error prone.

**Comment by enj** [comment]
deepcopy/defaulting did not change at all (seems correct based on how the current generated code is written).

We also no longer have the scripts to update the swagger descriptions for these types (I guess that does not matter since they do not serve REST endpoints anyway).

**Comment by enj** [comment]
Green tests but no proof any of it works!  :shipit:

**Comment by simo5** [comment]
/lgtm
needs to lift the other blocking flag before it can merge, but it is good enough for my taste
@liggitt feel free to cancel if you think the nitpicks of yours that went unaddressed are really important

**Comment by enj** [comment]
/hold cancel

@liggitt anything you want changed?

File: pkg/oauthserver/oauth/external/github/github.go (hunk 3 above)
Comment (liggitt) [reply pkg/oauthserver/oauth/external/github/github.go]: it does save you from setting the field in one place, not realizing you needed to override it in the other, and accidentally sending a token to github

File: pkg/oauthserver/oauth/external/github/github_test.go [hunk 6]
```diff
@@ -1,11 +1,29 @@
...
+	// set up the provider structs in a way that all config string values are different
+	standard := NewProvider("a", "b", "c", "", nil, nil, nil)
+	enterprise := NewProvider("d", "e", "f", "g", nil, nil, nil)
+
+	// get the raw structs using reflection
+	standardValue := reflect.ValueOf(standard).Elem()
+	enterpriseValue := reflect.ValueOf(enterprise).Elem()
+
+	// check to make sure all string fields are different
+	// any fields that are the same at this point are likely to be API endpoints we forgot to update
```
[the remark this answers is not in the document]
Comment (liggitt) [reply pkg/oauthserver/oauth/external/github/github_test.go]: I still think this is a weird test

**Comment by liggitt** [comment]
> @liggitt anything you want changed?

https://github.com/openshift/origin/pull/19493/files#r183873559 still seems error-prone

server config change looks fine

**Comment by enj** [comment]
Comments addressed.  @kalexand-rh this will also need docs for the two new fields.

**Comment by kalexand-rh** [comment]
Thanks @enj! The doc card's [here](https://trello.com/c/r6jwfTKr/884-document-github-enterprise-as-auth-provider), and we'll pick it up for 3.11.

**Pull request merged**